R/data-converters.R

```diff
@@ -0,0 +1,306 @@
+#' of [enw_preprocess_data()] and other [epinowcast()] preprocessing functions.
+#'
+#' @param linelist An object coercible to a `data.table` (such as a
+#' `data.frame`) where each row represents a case. Must contain at least
+#' two date variables or variables that can be coerced to dates.
+#'
+#' @param reference_date A date or a variable that can be coerced to a date
+#' that represents the date of interest for the case. For example, if the
+#' `reference_date` is the date of symptom onset then the `new_confirm` will
+#' be the number of new cases reported (based on `report_date`) on each day
+#' that had onset on that day. The default is "reference_date".
+#'
+#' @param report_date A date or a variable that can be coerced to a date
+#' that represents the date the case was reported. The default is "report_date".
+#'
+#' @param by A character vector of variables to also aggregate by (i.e. as well
+#' as using the `reference_date` and `report_date`). If not supplied
+#' then the function will aggregate by just the `reference_date` and
+#' `report_date`.
+#'
+#' @param max_delay The maximum number of days between the `reference_date`
+#' and the `report_date`. If not supplied then the function will use the
+#' maximum number of days between the `reference_date` and the `report_date`
+#' in the `linelist`. If the `max_delay` is less than the maximum number of
+#' days between the `reference_date` and the `report_date` in the `linelist`
+#' then the function will use this value instead and inform the user.
+#'
+#' @inheritParams enw_complete_dates
+#' @inheritParams enw_add_incidence
+#'
+#' @return A `data.table` with the following variables: `reference_date`,
+#' `report_date`, `new_confirm`, `confirm`, `delay`, and
+#' any variables specified in `by`.
+#'
+#' @family dataconverters
+#' @export
+#' @importFrom data.table as.data.table setkeyv
+#' @examples
+#' linelist <- data.frame(
+#'   onset_date = as.Date(c("2021-01-02", "2021-01-03", "2021-01-02")),
+#'   report_date = as.Date(c("2021-01-03", "2021-01-05", "2021-01-04"))
+#' )
+#' enw_linelist_to_incidence(linelist, reference_date = "onset_date")
+#'
+#' # Specify a custom maximum delay and allow completion beyond the maximum
+#' # observed delay
+#' enw_linelist_to_incidence(
+#'  linelist, reference_date = "onset_date", max_delay = 5,
+#'  completion_beyond_max_report = TRUE
+#' )
+enw_linelist_to_incidence <- function(linelist,
+                                      reference_date = "reference_date",
+                                      report_date = "report_date",
+                                      by = NULL, max_delay,
+                                      completion_beyond_max_report = FALSE,
+                                      copy = TRUE) {
+  counts <- coerce_dt(
+    linelist, required_cols = c(by, reference_date, report_date), copy = copy
+  )
+  data.table::setnames(counts, reference_date, "reference_date")
+  data.table::setnames(counts, report_date, "report_date")
```

Comment:
this is thinking ahead to an internal constant for these:
```suggestion
  data.table::setnames(
    counts,
    c(reference_date, report_date),
    c("reference_date", "report_date")
  )
```